sdk/compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/SdkInstall.hs

```diff
@@ -0,0 +1,254 @@
+releaseVersionFromLspId (LSP.IdString lspIdStr) = T.stripSuffix "-sdk-install-request" lspIdStr >>= eitherToMaybe . parseUnresolvedVersion
+releaseVersionFromLspId _ = Nothing
+
+handleShowMessageResponse :: MultiIdeState -> LSP.LspId 'LSP.WindowShowMessageRequest -> Either LSP.ResponseError (Maybe LSP.MessageActionItem) -> IO ()
+handleShowMessageResponse miState (releaseVersionFromLspId -> Just ver) res = do
+  installDatas <- atomically $ takeTMVar $ misSdkInstallDatasVar miState
+  let installData = getSdkInstallData ver installDatas
+      changeSdkStatus :: LSP.DiagnosticSeverity -> T.Text -> SdkInstallStatus -> IO ()
+      changeSdkStatus severity message newStatus = do
+        installDatas' <- updateSdkStatus miState installDatas ver severity message newStatus
+        atomically $ putTMVar (misSdkInstallDatasVar miState) installDatas'
+      disableSdk = changeSdkStatus LSP.DsError (missingSdkIdeDiagnosticMessage ver) SISDenied
+
+  case (sidStatus installData, res) of
+    (SISAsking, Right (Just (LSP.MessageActionItem (T.stripPrefix "Install SDK" -> Just _)))) -> do
+      -- Install accepted, start install process
+      installThread <- async $ do
+        setupSdkInstallReporter miState ver
+        outputLogVar <- newMVar ""
+        res <- tryForwardAsync $ installSdk ver outputLogVar $ updateSdkInstallReporter miState ver
+        handleInstallResult miState ver outputLogVar $ either Just (const Nothing) res
+        finishSdkInstallReporter miState ver
+      changeSdkStatus LSP.DsInfo (installingSdkIdeDiagnosticMessage ver) (SISInstalling installThread)
+    (SISAsking, _) -> disableSdk
+    (_, _) -> atomically $ putTMVar (misSdkInstallDatasVar miState) installDatas
+handleShowMessageResponse _ _ _ = pure ()
+
+-- try @SomeException that doesn't catch AsyncCancelled exceptions
+tryForwardAsync :: IO a -> IO (Either SomeException a)
+tryForwardAsync = tryJust @SomeException $ \case
+  (fromException -> Just AsyncCancelled) -> Nothing
+  e -> Just e
+
+handleSdkInstallCancelled :: MultiIdeState -> LSP.NotificationMessage 'LSP.CustomMethod -> IO ()
+handleSdkInstallCancelled miState notif = do
+  forM_ (fromJSON $ notif ^. LSP.params) $ \message -> do
+    let ver = sicSdkVersion message
+    installDatas <- atomically $ takeTMVar $ misSdkInstallDatasVar miState
+    let installData = getSdkInstallData ver installDatas
+    installDatas' <- case sidStatus installData of
+      SISInstalling thread -> do
+        logDebug miState $ "Killing install thread for " <> unresolvedReleaseVersionToString ver
+        cancel thread
+        updateSdkStatus miState installDatas ver LSP.DsError (missingSdkIdeDiagnosticMessage ver) SISDenied
+      _ -> pure installDatas
+    atomically $ putTMVar (misSdkInstallDatasVar miState) installDatas'
+
+handleInstallResult :: MultiIdeState -> UnresolvedReleaseVersion -> MVar T.Text -> Maybe SomeException -> IO ()
+handleInstallResult miState ver outputLogVar mError = do
```

Comment:
In general the names in this module are a bit vague - I'm not always sure whether it's something from the client being handled or something the server is sending out to the client, or some stateful change. Can we make it more clear which functions are part of the client<-> server ShowMessage dance and which parts are more directly related to installation?

Reply:
I've done some renaming and comments, hopefully its clearer now